backoff.go

```diff
@@ -0,0 +1,132 @@
+
+// BackoffStrategy is a function for computing delays between attempts to
+// perform some application-defined operation.
+//
+// n is the number of successive failures since the last success, including this
+// one.
+//
+// err is the error describing the operation's failure, if known. A nil error
+// does not indicate a success.
+type BackoffStrategy func(n int, err error) time.Duration
+
+// DefaultBackoffStrategy is the default strategy used by Backoff.
+//
+// It is a conservative policy favouring large delay times under the assumption
+// that the operation is expensive.
+var DefaultBackoffStrategy BackoffStrategy = ExponentialBackoff(3 * time.Second)
+
+// Backoff introduces delays between atempts to perform some application-defined
+// operation.
+type Backoff struct {
+	// Strategy is the backoff strategy used to compute the "fundamental wait
+	// duration". If it is nil, DefaultBackoffStrategy is used.
+	Strategy BackoffStrategy
+
+	// Transform is applied to the "fundamental wait duration" to produce the
+	// "transformed wait duration".
+	//
+	// If it is nil, FullJitter is used. The Identity transform can be used to
+	// disable the transform.
+	Transform DurationTransform
+
+	// Min is a lower bound for the "transformed wait duration", in combination
+	// with Max it is used to produce the "bounded wait duration".
+	Min time.Duration
+
+	// Min is a upper bound for the "transformed wait duration", in combination
+	// with Min it is used to produce the "bounded wait duration".
+	Max time.Duration
+
+	// failures is the number of successive failures that have occurred.
+	failures uint32
+}
+
+// Ok marks the most recent attempt as a success.
+func (b *Backoff) Ok() {
+	atomic.StoreUint32(&b.failures, 0)
+}
+
+// Fail marks the most recent attempt as a failure and returns the "bounded wait
+// duration", which is the time to wait before retrying the operation.
+//
+// err is the error describing the operation's failure, if known. A nil error
+// does not indicate a success.
+func (b *Backoff) Fail(err error) time.Duration {
+	min := Longest(b.Min, 0)
+	max := MustCoalesce(b.Max, 1*time.Hour)
+
+	strategy := b.Strategy
+	if strategy == nil {
+		strategy = DefaultBackoffStrategy
+	}
+
+	transform := b.Transform
+	if transform == nil {
+		transform = FullJitter
+	}
+
+	n := atomic.AddUint32(&b.failures, 1)
+
+	d := strategy(int(n), err) // fundamental
+	d = transform(d)           // transformed
+	d = Limit(d, min, max)     // bounded
+
+	return d
+}
+
+// Sleep marks the most recent attempt and pauses the current goroutine until
+// the "bounded wait duration" has elapsed.
+//
+// It sleeps until the duration elapses or ctx is canceled, whichever is first.
+// If ctx is canceled before the duration elapses it returns ctx.Err(),
+// otherwise it returns nil.
+//
+// err is the error describing the operation's failure, if known. A nil error
+// does not indicate a success.
+func (b *Backoff) Sleep(ctx context.Context, err error) error {
```

Comment:
Is this a recursive call? (Well, definitely not recursive, sorry for being such a ...)

Reply:
I can't seem to find Sleep() function anywhere. Is it in another file?

Reply:
Yeah, it's the existing `Sleep()` function from this library.  It sleeps for a duration OR until `ctx` is done.

Reply:
I would have thought you could click on it to go-to definition, I guess that doesn't work in PRs.